docs/operator.md

```diff
@@ -0,0 +1,117 @@
+# Operator build and deployment
+
+## Prerequisites
+
+1. operator-sdk <https://sdk.operatorframework.io/docs/installation/>
+2. kustomize <https://github.com/kubernetes-sigs/kustomize/releases>
+3. opm <https://github.com/operator-framework/operator-registry/releases>
+
+## Building the operator
+
+### Create the bundle manifests
+
+```bash
+make bundle
+```
+
+To generate the manifests and CSV in ./bundle, "make bundle" first calls the "ocp-create-manifests" target. This target in turn calls "deploy-service-on-ocp-cluster and deploy-ui-on-ocp-cluster" while setting APPLY_MANIFESTS=False and APPLY_NAMESPACE=False. This causes the resource yamls to be created in ./build/assisted-installer/ but does not apply them.
+
+The relevant resource yamls are then copied to ./config/assisted-service where additional customizations are applied using Kustomize. The resulting yaml is then piped to operator-sdk
+creating the manifests and CSVs in ./bundle/manifests. 
+
+More information about bundles: <https://sdk.operatorframework.io/docs/olm-integration/generation/>
+
+### Create the bundle and index images
+
+```bash
+export ORG=quay.io/ocpmetal
+make bundle-build BUNDLE_IMG=$ORG/assisted-service-operator-bundle:0.0.1
+podman push $ORG/assisted-service-operator-bundle:0.0.1
+
+opm index add --bundles $ORG/assisted-service-operator-bundle:0.0.1 --tag $ORG/assisted-service-index:0.0.1
+podman push $ORG/assisted-service-index:0.0.1
+```
+
+## Deploying the operator through OperatorHub
+
+The operator must be deployed to assisted-installer namespace. Create the namespace. 
+
+```bash
+cat <<EOF | kubectl create -f -
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: assisted-installer
+  labels:
+    name: assisted-installer
+EOF
+```
+
+Create two PVCs, one for postgres and another for s3 (temporary until filesystem implementation is available).
```

Comment:
Why don't we create the PVC as a part of the deployment?
Also I don't see the volume mount in the PG deployment, how does it end up using this?

Reply:
For the first question, unfortunately PVC can't be specified in the operator's manifests https://docs.openshift.com/container-platform/4.6/operators/understanding/olm-packaging-format.html

Reply:
So that says:
> When these optional objects are included in a bundle, Operator Lifecycle Manager (OLM) can create them from the bundle and manage their lifecycle along with the CSV

That doesn't exactly mean that we *can't* create other objects as a part of the manifest, just that they won't follow the operator's lifecycle. Which in the case of a PVC is probably good as removing it would cause data loss.
If we do include it in the spec somewhere does some part of this process (bundling or installing, presumably) error out?

Reply:
I'm intrigued :thinking: need to think about this.

Reply:
You're right, the deployment doesn't define a volume mount, so the PVC ends up being unused. Since we are intending to support persistence, the volume mount will need to be added..

Reply:
I tested adding PVC to the operator bundle and it makes the operator undeployable. Operator becomes stuck at RequirementsNotMet status and olm-operator logs doesn't show what the problem could be.